File: core/byosnap-node-ts/src/controllers/settingsController.ts
import { ErrorResponse, SuccessMessageResponse } from '../types/responses';
import { SettingsSchema, getDefaultSettings } from '../models/settingsModel';
import { ExportSettingsSchema } from '../models/exportSettingsModel';
import { CustomSettingsPayload } from '../models/customSettingsModel';
import { authMiddleware } from '../middleware/authMiddleware';
import { Request as ExpressRequest } from 'express';
import { Controller, Route, Get, Put, Post, Delete, Path, Query, Extension, Body, Middlewares, TsoaResponse, Response, Res, Request } from 'tsoa';

// # @GOTCHAS 👋 - Please read GOTCHAS.md

// ===========================================================================
// Settings Controller — System endpoints Snapser expects every BYOSnap to
// expose (configuration tools, import/export, user manager, GDPR).
//
// This is the CORE starter scaffold: every endpoint is present but each handler
// body is a STUB that returns a simple placeholder. Fill in the stubs with your
// own logic. For a complete, working reference (Storage Snap reads/writes, CAS
// handling, validation) see advanced/byosnap-node-ts.
// ===========================================================================

@Route('/v1/byosnap-core/settings')
export class SettingsController extends Controller {

    // ===========================================================================
    // A i]: Configuration Tool: Built using the Snapser UI Builder
    // ===========================================================================

    /**
     * @summary Configuration Tool
     */
    @Get("/")
    @Extension("x-description", 'Get the settings for this Snap. This endpoint is called by the Snapser Configuration Tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Extension("x-snapser-sdk-categories", ["admin"])
    @Response<SettingsSchema>(200, "Settings retrieved successfully")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async getSettings(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Query() tool_id?: string,
        @Query() environment?: string,
    ): Promise<SettingsSchema> {
        // TODO: Fetch the saved settings for `${tool_id}_${environment}` (e.g. from
        //       the Storage Snap) and return them. For now we return the default.
        //       See advanced/byosnap-node-ts for the full implementation.
        return getDefaultSettings();
    }

    /**
     * @summary Configuration Tool
     */
    @Put("/")
    @Extension("x-description", 'Update the settings for this Snap. This endpoint is called by the Snapser Configuration Tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Extension("x-snapser-sdk-categories", ["admin"])
    @Response<SettingsSchema>(200, "Settings updated successfully")
    @Response<ErrorResponse>(400, "Bad request")
    @Response<ErrorResponse>(500, "Server error")
    @Middlewares([authMiddleware(["internal"])])
    public async updateSettings(
        @Res() _badRequest: TsoaResponse<400, ErrorResponse>,
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Body() body: Record<string, any>,
        @Query() tool_id?: string,
        @Query() environment?: string,
    ): Promise<Record<string, any>> {
        // TODO: Validate the incoming settings and persist them for
        //       `${tool_id}_${environment}`. On a validation failure return:
        //       return _badRequest(400, { error_message: '...' });
        //       See advanced/byosnap-node-ts for the full implementation.
        return { message: 'Success' };
    }

    // ===========================================================================
    // A ii]: New Configuration Tool: Custom HTML Snap Configuration Tool
    // ===========================================================================

    /**
     * @summary Custom Configuration Tool
     */
    @Get("custom")
    @Extension("x-description", 'Get the settings for the custom HTML configuration tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Extension("x-snapser-sdk-categories", ["admin"])
    @Response<CustomSettingsPayload>(200, "Custom settings retrieved successfully")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async getSettingsCustom(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Query() tool_id?: string,
        @Query() environment?: string,
    ): Promise<CustomSettingsPayload> {
        // TODO: Fetch the saved settings for `${tool_id}_${environment}` and return
        //       them wrapped as { payload: <settings> }.
        //       See advanced/byosnap-node-ts for the full implementation.
        return { payload: '' };
    }

    /**
     * @summary Custom Configuration Tool
     */
    @Put("custom")
    @Extension("x-description", 'Update the settings from the custom HTML configuration tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Extension("x-snapser-sdk-categories", ["admin"])
    @Response(200, "Custom settings updated successfully")
    @Response<ErrorResponse>(500, "Server error")
    @Middlewares([authMiddleware(["internal"])])
    public async updateSettingsCustom(
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Body() body: Record<string, any>,
        @Query() tool_id?: string,
        @Query() environment?: string,
    ): Promise<Record<string, any>> {
        // TODO: Validate the incoming settings and persist them for
        //       `${tool_id}_${environment}`.
        //       See advanced/byosnap-node-ts for the full implementation.
        return { message: 'Success' };
    }

    // ===========================================================================
    // B: Snapend Sync|Clone: Import/Export
    // ===========================================================================

    /**
     * @summary Export Settings
     */
    @Get("export")
    @Extension("x-description", 'Export all settings across environments (dev, stage, prod) for Snapend Sync/Clone.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<ExportSettingsSchema>(200, "Settings exported successfully")
    @Response<ErrorResponse>(500, "Server error")
    @Middlewares([authMiddleware(["internal"])])
    public async exportSettings(
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Request() req: ExpressRequest,
    ): Promise<ExportSettingsSchema> {
        // TODO: Load the real saved settings per environment and merge them into
        //       `data`. For now we return defaults for dev/stage/prod.
        //       See advanced/byosnap-node-ts for the full implementation.
        return {
            version: process.env.BYOSNAP_VERSION ?? '1.0.0',
            exported_at: Math.floor(Date.now() / 1000),
            data: {
                dev: { characters: getDefaultSettings() },
                stage: { characters: getDefaultSettings() },
                prod: { characters: getDefaultSettings() },
            },
        };
    }

    /**
     * @summary Import Settings
     */
    @Post("import")
    @Extension("x-description", 'Import settings across environments (dev, stage, prod) for Snapend Sync/Clone.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<SuccessMessageResponse>(200, "Settings imported successfully")
    @Response<ErrorResponse>(400, "Bad request")
    @Response<ErrorResponse>(500, "Server error")
    @Middlewares([authMiddleware(["internal"])])
    public async importSettings(
        @Res() _badRequest: TsoaResponse<400, ErrorResponse>,
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Body() body: ExportSettingsSchema,
    ): Promise<SuccessMessageResponse> {
        // TODO: Validate the incoming export payload and persist each environment's
        //       settings. On a validation failure return:
        //       return _serverError(500, { error_message: 'Invalid JSON' });
        //       See advanced/byosnap-node-ts for the full implementation.
        return { message: 'Success' };
    }

    /**
     * @summary Validate Import Settings
     */
    @Post("validate-import")
    @Extension("x-description", 'Validate settings before importing. Snapser sends the settings that are about to be imported - validate if you can accept them.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<ExportSettingsSchema>(200, "Settings are valid")
    @Response<ErrorResponse>(500, "Invalid settings")
    @Middlewares([authMiddleware(["internal"])])
    public async validateImportSettings(
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Request() req: ExpressRequest,
        @Body() body: ExportSettingsSchema,
    ): Promise<ExportSettingsSchema> {
        // TODO: Add your own validation here. On failure return:
        //       return _serverError(500, { error_message: 'Invalid JSON' });
        //       See advanced/byosnap-node-ts for the full implementation.
        return body;
    }

    // ===========================================================================
    // A iii]: User Manager Tool: Custom HTML User Manager Tool
    // ===========================================================================

    /**
     * @summary User Manager Tool
     */
    @Get("users/{userId}/custom")
    @Extension("x-description", 'Get the user data for the custom HTML User Manager tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response<CustomSettingsPayload>(200, "User data retrieved successfully")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async getUserDataCustom(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<CustomSettingsPayload> {
        // TODO: Look up this user's data and return it wrapped as
        //       { payload: <data> }.
        //       See advanced/byosnap-node-ts for the full implementation.
        return { payload: '' };
    }

    /**
     * @summary User Manager Tool
     */
    @Post("users/{userId}/custom")
    @Extension("x-description", 'Update the user data for the custom HTML User Manager tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response(200, "User data updated successfully")
    @Response<ErrorResponse>(500, "Server error")
    @Middlewares([authMiddleware(["internal"])])
    public async updateUserDataCustom(
        @Res() _serverError: TsoaResponse<500, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
        @Body() body: Record<string, any>,
    ): Promise<Record<string, any>> {
        // TODO: Validate the incoming data and persist it for this userId.
        //       See advanced/byosnap-node-ts for the full implementation.
        return { message: 'Success' };
    }

    // ===========================================================================
    // C: User Tool: Get, Update and Delete User data: GDPR
    // ===========================================================================

    /**
     * @summary User Data (GDPR)
     */
    @Get("users/{userId}/data")
    @Extension("x-description", 'Get user data. Used by the GDPR tool and the User Manager tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response(200, "User data retrieved successfully")
    @Response<ErrorResponse>(400, "No data found")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async getUserData(
        @Res() _badRequest: TsoaResponse<400, ErrorResponse>,
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<Record<string, any>> {
        // TODO: Fetch and return everything you store for this userId.
        //       See advanced/byosnap-node-ts for the full implementation.
        return {};
    }

    /**
     * @summary User Data (GDPR)
     */
    @Put("users/{userId}/data")
    @Extension("x-description", 'Update user data. Used by the GDPR tool and the User Manager tool.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response(200, "User data updated successfully")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async updateUserData(
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<Record<string, any>> {
        // TODO: Persist the incoming data for this userId.
        //       See advanced/byosnap-node-ts for the full implementation.
        return {};
    }

    /**
     * @summary User Data (GDPR)
     */
    @Delete("users/{userId}/data")
    @Extension("x-description", 'Delete user data. Implements the GDPR right-to-be-forgotten.')
    @Extension("x-snapser-auth-types", ["internal"])
    @Response(200, "User data deleted successfully")
    @Response<ErrorResponse>(400, "No blob found")
    @Response<ErrorResponse>(401, "Unauthorized")
    @Middlewares([authMiddleware(["internal"])])
    public async deleteUserData(
        @Res() _badRequest: TsoaResponse<400, ErrorResponse>,
        @Res() _unauthorized: TsoaResponse<401, ErrorResponse>,
        @Path() userId: string,
        @Request() req: ExpressRequest,
    ): Promise<Record<string, any>> {
        // TODO: Delete everything you store for this userId.
        //       See advanced/byosnap-node-ts for the full implementation.
        return {};
    }
}

File: core/byosnap-java/src/main/java/com/snapser/byosnap/controller/SettingsController.java
package com.snapser.byosnap.controller;

import java.util.List;
import java.util.Map;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import com.snapser.byosnap.AppConstants;
import com.snapser.byosnap.auth.ValidateAuthorization;

/**
 * System endpoints Snapser expects every BYOSnap to expose.
 *
 * <p>These power the Configuration Tool (UI Builder + Custom HTML), the User
 * Manager Tool, the Snapend Sync/Clone import-export system, and the GDPR user
 * data tooling. Every handler here is guarded {@code internal} because Snapser
 * calls them through the internal gateway. Each body is a STUB — fill in the
 * TODOs with your own persistence (e.g. via the Snapser Storage snap).
 */
@RestController
public class SettingsController {

    // =====================================================================
    // A i] Configuration Tool: built with the Snapser UI Builder
    // =====================================================================

    @GetMapping(AppConstants.API_PREFIX + "/settings")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> getSettings(
            @RequestParam(name = "tool_id", required = false) String toolId,
            @RequestParam(name = "environment", required = false, defaultValue = AppConstants.DEFAULT_ENVIRONMENT) String environment) {
        // This is the default payload shape the Configuration Tool expects.
        Map<String, Object> defaultSettings = Map.of(
                "sections", List.of(Map.of(
                        "id", "registration",
                        "components", List.of(Map.of(
                                "id", "characters",
                                "type", "textarea",
                                "value", "")))));
        String blobOwnerKey = toolId + "_" + environment;
        // TODO: Fetch the saved settings for `blobOwnerKey` (e.g. from the Storage
        //       snap) and return them. For now we return the default shape.
        return ResponseEntity.ok(defaultSettings);
    }

    @PutMapping(AppConstants.API_PREFIX + "/settings")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> updateSettings(
            @RequestParam(name = "tool_id", required = false) String toolId,
            @RequestParam(name = "environment", required = false, defaultValue = AppConstants.DEFAULT_ENVIRONMENT) String environment,
            @RequestBody(required = false) Map<String, Object> body) {
        String blobOwnerKey = toolId + "_" + environment;
        Object blobData = (body != null && body.containsKey("payload")) ? body.get("payload") : body;
        // TODO: Validate `blobData` and persist it for `blobOwnerKey` (e.g. via
        //       the Storage snap). On a validation failure return a 400 with an
        //       error_message body.
        return ResponseEntity.ok(Map.of("success", true));
    }

    // =====================================================================
    // A ii] Configuration Tool: custom HTML Snap configuration tool
    // =====================================================================

    @GetMapping(AppConstants.API_PREFIX + "/settings/custom")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> getSettingsCustom(
            @RequestParam(name = "tool_id", required = false) String toolId,
            @RequestParam(name = "environment", required = false, defaultValue = AppConstants.DEFAULT_ENVIRONMENT) String environment) {
        String blobOwnerKey = toolId + "_" + environment;
        // TODO: Fetch the saved settings for `blobOwnerKey` and return them
        //       wrapped as {"payload": <settings>}.
        return ResponseEntity.ok(Map.of("payload", ""));
    }

    @PutMapping(AppConstants.API_PREFIX + "/settings/custom")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> updateSettingsCustom(
            @RequestParam(name = "tool_id", required = false) String toolId,
            @RequestParam(name = "environment", required = false, defaultValue = AppConstants.DEFAULT_ENVIRONMENT) String environment,
            @RequestBody(required = false) Map<String, Object> body) {
        String blobOwnerKey = toolId + "_" + environment;
        Object blobData = (body != null && body.containsKey("payload")) ? body.get("payload") : body;
        // TODO: Validate `blobData` and persist it for `blobOwnerKey`. On a
        //       validation failure return a 400 with an error_message body.
        return ResponseEntity.ok(Map.of("success", true));
    }

    // =====================================================================
    // A iii] User Manager Tool: custom HTML user manager tool
    // =====================================================================

    @GetMapping(AppConstants.API_PREFIX + "/settings/users/{userId}/custom")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> getUserDataCustom(@PathVariable String userId) {
        // TODO: Look up this user's data (e.g. from the Storage snap) and return
        //       it wrapped as {"payload": <data>}.
        return ResponseEntity.ok(Map.of("payload", ""));
    }

    @PostMapping(AppConstants.API_PREFIX + "/settings/users/{userId}/custom")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> updateUserDataCustom(
            @PathVariable String userId,
            @RequestBody(required = false) Map<String, Object> body) {
        Object blobData = (body != null && body.containsKey("payload")) ? body.get("payload") : body;
        // TODO: Validate `blobData` and persist it for this userId.
        return ResponseEntity.ok(Map.of("success", true));
    }

    // =====================================================================
    // B] Snapend Sync/Clone: Snapser's built-in configuration import/export
    // =====================================================================

    @GetMapping(AppConstants.API_PREFIX + "/settings/export")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> settingsExport() {
        // Snapser calls this when cloning/syncing a Snapend. Return your settings
        // for every environment so they can be re-imported elsewhere.
        Map<String, Object> emptyToolPayload = Map.of(
                "sections", List.of(Map.of(
                        "id", "registration",
                        "components", List.of(Map.of(
                                "id", "characters",
                                "type", "textarea",
                                "value", "")))));
        String version = System.getenv().getOrDefault(
                AppConstants.BYOSNAP_VERSION_ENV_KEY, AppConstants.DEFAULT_BYOSNAP_VERSION);
        Map<String, Object> response = Map.of(
                "version", version,
                "exported_at", 0,
                "data", Map.of(
                        // The key here ("characters") is the Tool ID whose payload you export.
                        "dev", Map.of("characters", emptyToolPayload),
                        "stage", Map.of("characters", emptyToolPayload),
                        "prod", Map.of("characters", emptyToolPayload)));
        // TODO: Load the real saved settings per environment (e.g. batch-get from
        //       the Storage snap) and merge them into response["data"].
        return ResponseEntity.ok(response);
    }

    @PostMapping(AppConstants.API_PREFIX + "/settings/import")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> settingsImport(@RequestBody(required = false) Map<String, Object> body) {
        // TODO: Validate the incoming export payload and persist each
        //       environment's settings (e.g. batch-replace on the Storage snap).
        if (body == null || body.isEmpty()) {
            return ResponseEntity.internalServerError().body(Map.of("error_message", "Invalid JSON"));
        }
        return ResponseEntity.ok(Map.of("message", "Success"));
    }

    @PostMapping(AppConstants.API_PREFIX + "/settings/validate-import")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> validateSettingsImport(@RequestBody(required = false) Map<String, Object> body) {
        // Snapser sends the settings it is about to import. Decide whether you
        // can accept them: return 200 with the payload if so, 500 otherwise.
        // TODO: Add your own validation here (e.g. check the dev/stage/prod
        //       structure).
        if (body == null || body.isEmpty()) {
            return ResponseEntity.internalServerError().body(Map.of("error_message", "Invalid JSON"));
        }
        return ResponseEntity.ok(body);
    }

    // =====================================================================
    // C] User Tool: get / update / delete user data
    //    Used by the GDPR tool and the User Manager tool.
    // =====================================================================

    @GetMapping(AppConstants.API_PREFIX + "/settings/users/{userId}/data")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> getUserData(@PathVariable String userId) {
        // TODO: Fetch and return everything you store for this userId (e.g. from
        //       the Storage snap).
        return ResponseEntity.ok(Map.of());
    }

    @PutMapping(AppConstants.API_PREFIX + "/settings/users/{userId}/data")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> updateUserData(
            @PathVariable String userId,
            @RequestBody(required = false) Map<String, Object> body) {
        // TODO: Persist the incoming data for this userId.
        return ResponseEntity.ok(Map.of());
    }

    @DeleteMapping(AppConstants.API_PREFIX + "/settings/users/{userId}/data")
    @ValidateAuthorization(AppConstants.GATEWAY_INTERNAL_ORIGIN)
    public ResponseEntity<Object> deleteUserData(@PathVariable String userId) {
        // TODO: Delete everything you store for this userId (right-to-be-forgotten).
        return ResponseEntity.ok(Map.of());
    }
}
